Implement Segwit for BSQ

stejbac opened the pull request:
<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

This PR attempts to add support for native Segwit (p2wpkh) BSQ addresses.

The [btcd-cli4j](https://github.com/priiduneemre/btcd-cli4j) Json-RPC interface for bitcoind (which is used by DAO full nodes) is replaced, as that code appears to be unmaintained for about 5 years and only maintained compatibility up to bitcoind-0.10.3. In particular, it does not support the extra Segwit fields in the `getblock` Json responses returned to the Java client, including the `txinwitness` field of the raw tx inputs. These are needed to fill in the `TxInput.pubKey` fields of the inputs of each BSQ tx that makes up the DAO state, used in turn to validate staked merit (from issuance txs) and messages signed for a proof-of-burn tx. Thus the RPC client needs upgrading for us to support compensation and proof-of-burn txs with native Segwit BSQ inputs.

A new, home-grown RPC client is provided using the [jsonrpc4j](https://github.com/briandilley/jsonrpc4j) library. Additionally, a simple home-grown socket daemon is provided to receive block hashes sent by the _blocknotify_ script run by bitcoind, replacing the btcd-cli4j version.

- TODO: Improve the RPC client startup logic - should send a ping or `getnetworkinfo` message to bitcoind & check the version. **[edit: now done]**
- TODO: Fix a consensus issue caused by failure of _RpcService.java_ to account for at least one spurious fully native Segwit BSQ tx (d1f45e55be6101b1b75e6bf9fc5e5341c6ab420647be7555863bbbddd84e92f3) which appeared in block 660384, dated 2020-12-07 (thus missing a `TxInput.pubKey` field in the DAO state). I'm not sure yet if there are any other consensus issues. **[edit: now done]**
- TODO: Actually implement and enable the native segwit BSQ keychain and make sure there are no issues.
  **[edit: should be finished - created a separate PR #5109 for the changes]**

stejbac committed:
Bring ScryptType enum up to date with latest bitcoind

Add 'witness_v1_taproot' script type to the enum and proto.pb, so that
it doesn't cause any problems when Taproot is activated and the new
script type starts showing up in RPC getBlock(..) responses (including
possibly BSQ transactions).

Also change the Java enum order (which shouldn't cause any problems as
the ordinal isn't used directly in hashCode calculations) and add the
missing 'witness_unknown' enum value to pb.proto to bring it in sync.

stejbac committed:
Add replacement bitcoind RPC client using jsonrpc4j

Create a new 'BitcoindClient' interface and a corresponding builder, to
replace the old 'com.neemre.btcdcli4j.core.client.BtcdClientImpl' class
from the btcdcli4j library. This is instantiated by jsonrpc4j using a
dynamic proxy. It provides only a cut down version of the bitcoind RPC
API, exposing the methods 'getblock', 'getblockcount' & 'getblockhash',
as they are the only ones currently being used by RpcService.

Add corresponding Jackson-annotated DTO classes to model the JSON
structures returned by bitcoind, very similar to the classes provided by
btcdcli4j. Note that we use Double instead of BigDecimal to represent
fractional fields (difficulties + coin amounts in BTC), as they have
more consistent Jackson (de)serialisation and appear to be able to
faithfully round-trip numeric fields produced by bitcoind. Also note
that doubles can faithfully represent any valid decimal BTC amount (that
is, with 8 d.p. of precision) up to 21 million.

For now, keep the old BtcdClientImpl instance used by RpcService in
place, as the btcdcli4j block notification daemon is dependent upon it
and would also need to be replaced.

Also add unit tests for BitcoindClient which test against sample regtest
responses, using a mock HttpURLConnection.

stejbac committed:
Add replacement bitcoind block notification daemon

Provide a new 'BitcoindDaemon' block notification socket server, to
replace 'com.neemre.btcdcli4j.daemon.BtcdDaemonImpl'. This starts a
single service thread to listen for raw block hashes on localhost port
512*, sent by the specified 'blocknotify' shell/batch script, delegating
to a pool of worker threads to run the supplied BlockListener handler.
Unlike the original BtcdDaemonImpl class, a call to the 'getblock' RPC
method is not made automatically to supply a complete block to the
handler, instead requiring a separate, manual BitcoindClient.getBlock
invocation from within RpcService.

Also provide unit tests using a mock ServerSocket + Socket.

TODO: Use the new Bitcoind(Client|Daemon) implementations in RpcService,
 in place of btcdcli4j Btcd(Client|Daemon)Impl & remove the old library.

stejbac committed:
Factor out executor shutdown logic into new Utilities method

stejbac committed:
Factor out failure callback logic into new Utilities method

stejbac committed:
Add missing copyright headers to new files

stejbac committed:
Improve exception handling in BitcoindDaemon

Wrap any exception that occurs during socket IO or within the supplied
BlockListener with a new 'BlockNotificationException'. This brings the
exception handling more in line with that of the old BtcdDaemonImpl and
makes it easier to match them downstream in FullNode.handleError.

stejbac committed:
Use new Bitcoind(Client|Daemon) & remove btcd-cli4j

Migrate RpcService over to the new block notification daemon and RPC
client based on jsonrpc4j. Drop in own DTO classes in place of the ones
defined by btcd-cli4j and rename requestBtcBlock & addNewBtcBlockHandler
to requestDtoBlock & addNewDtoBlockHandler respectively.

Also remove now redundant filtering from the logback config and update
grade-witness.

stejbac committed:
Enable extraction of segwit pubkeys from raw tx inputs

Factor out a new RpcService.extractPubKeyAsHex method, to take public
keys from the inputs of the raw transactions returned by the RPC client,
when building TxInput objects to incorporate into the DAO state. Enhance
the method to additionally support segwit (P2WPKH & P2SH-P2WPKH) inputs
(but only the first input for backwards compatibility - see code
comment). Also fix a bug when handling non-SIGHASH_ALL input signatures.

This will allow segwit BSQ to be used in proof-of-burn and issuance txs,
which need a public key associated with the tx to establish ownership of
it, when signing messages with a proof-of-burn or staking merit awarded
from a compensation issuance, respectively.

Also add unit tests for the factored-out method and add a missing RawTx
toString() method, to aid debugging the TxInput fields within the
processed block returned by RpcService.

stejbac committed:
Fix Codacy issues & deduplicate code in RpcService

Factor out shared construction logic to a new 'getBlockFromRawDtoBlock'
method in RpcService. Also add some 'NOPMD' comments in an attempt to
suppress unfixable Codacy warnings about qualified imports.

chimp1984 commented:
Ah great to see work on that! I will have a look soon into it...

>     * TODO: Fix a consensus issue caused by failure of _RpcService.java_ to account for at least one spurious fully native Segwit BSQ tx (d1f45e55be6101b1b75e6bf9fc5e5341c6ab420647be7555863bbbddd84e92f3) which appeared in block 660384, dated 2020-12-07 (thus missing a `TxInput.pubKey` field in the DAO state). I'm not sure yet if there are any other consensus issues.

I assume some dev has created that tx. With the release code base its should not be possible to create a BSQ tx with segwit input.
The amount got burned:
https://mempool.emzy.de/de/bisq/tx/d1f45e55be6101b1b75e6bf9fc5e5341c6ab420647be7555863bbbddd84e92f3

chimp1984 commented:
I reviewed the current state. Looks all good to me. Will do a more in depth review once its out of draft state.

stejbac committed:
Fix data race in BitcoindDaemonTest

Prevent intermittent test failures, caused by a race between checking
whether the mock socket is closed upon accepting a new connection and
setting 'socketClosed' to true during shutdown. Waiting to accept and
then checking the flag needs to be done in a synchronized block.

stejbac committed:
Make special case for irregular tx with segwit BSQ inputs

Selectively disable pubkey extraction from segwit inputs of a particular
tx at block height 660384 (2020-12-07), which spends spuriously created
segwit BSQ (later burned), to prevent a change in the DAO state hashes
from that point.

(Since a tx with a given ID can only appear on one chain, a fixed global
exclusion list of IDs should not cause any issues on testnet/regtest
versus mainnet. This is simpler than conditioning by block height.)

chimp1984 commented on core/src/main/java/bisq/core/dao/node/full/RpcService.java:
```diff
@@ -246,14 +254,18 @@ private static RawTx getTxFromRawTransaction(RawTransaction rawDtoTx,
                     }
-                    return new TxInput(rawInput.getTxId(), rawInput.getVOut(), pubKeyAsHex);
+                    var txInput = new TxInput(rawInput.getTxId(), rawInput.getVOut(), pubKeyAsHex);
+                    txInput.isSegwit = rawInput.getTxInWitness() != null;
```
Wouldn't it be better to add that to the constructor so the TxInput is immutable?

stejbac replied:
(No longer applicable)

stejbac committed:
Add getNetworkInfo & getBestBlockHash RPC client methods

Provide a 'NetworkInfo' DTO class (with associated nested DTO classes),
returned by the 'getnetworkinfo' RPC method call to bitcoind. This will
be used during startup of RpcService to determine if Bitcoin Core is
available and which version it is using. Add a unit test to round-trip a
sample NetworkInfo JSON response.

Also add the missing 'getbestblockhash' RPC method, which will be needed
by RpcService to determine the Bitcoin Core node health.

stejbac committed:
Add Bitcoin Core version & health check to RpcService

Make a 'getnetworkinfo' RPC call to bitcoind immediately upon startup,
to check that the node is up (and throw a ConnectException to ensure
that the user is presented with an appropriate warning popup otherwise).
Log a warning if the node version is outside the 0.18.0 - 0.20.1 range.

Additionally, call 'getbestblockhash' to check that the chain tip is not
stale (> 6 hours old). As part of this, make sure the 'getblock' RPC
call works correctly with verbosity < 2, by fixing JSON deserialisation
of the response when the block or txs are in summary (hex string) form.

(These version & health checks are almost identical to the ones done by
the original btcd-cli4j library during RPC client startup.)

stejbac commented:
Rebasing onto a more recent version of master (00c82bb7 on 2020/12/28, 1.5.2 merge, #5012), to avoid conflicts in build.gradle and gradle-witness.gradle caused by a bitcoinj upgrade...

stejbac commented:
I'm thinking it may be better to put the actual BSQ wallet changes in a separate PR, so that anyone running a full DAO node has a chance to upgrade to a Bisq version including these RPC changes first, before segwit BSQ txs start appearing which would otherwise break their consensus with the rest of the network. (So the above changes should now be ready for review.)

chimp1984 commented:
Have you done a full DAO rebuild from genesis with that branch and check if the hashes in the monitor are the same as those from current seeds? It takes a while, probably several hours or so....

Have you seen my ealrier commnet at a review? ("Wouldn't it be better to add that to the constructor so the TxInput is immutable?")

stejbac commented:
I did a full DAO rebuild with an earlier version of the current branch. I should probably do another rebuild - I'll start that now.

TxInput should be immutable in the current version. There was an experimental `isSegwit` mutable flag that I added to it for testing and accidentally pushed that about a week ago, then amended and force pushed the intended commit (fdc739d) a little while later, so I think that dismissed your review comment (which I missed).

stejbac commented:
I've finished a DAO rebuild from genesis with the current branch and the hashes are still in agreement with the seed nodes.

chimp1984 commented:
@stejbac I think it would be more safe if we fork the jsonrpc4j libray to our own repo and use that. What do you think?

chimp1984 commented on core/src/main/java/bisq/core/dao/node/full/RpcService.java:
```diff
@@ -246,8 +254,10 @@ private static RawTx getTxFromRawTransaction(RawTransaction rawDtoTx,
+                    // recorded in the DAO state prior to the segwit upgrade of the RPC client. Spurious
+                    // segwit BSQ inputs in txs mined prior to the upgrade also require exclusion.
+                    String pubKeyAsHex = extractPubKeyAsHex(rawInput, rawInput == rawDtoTx.getVIn().get(0) &&
+                            !BSQ_TXS_DISALLOWING_SEGWIT_PUB_KEYS.contains(txId));
```
Wouldn't it be better to use a block height at which we activate the support for segwit inputs. It is technically a hard fork which would become activated at that block height.
That should also cover the past tx(s) with a segwit input.

stejbac replied:
Yes, perhaps that would be better. I'll push a change to do that shortly, as well as no longer excluding all but the first segwit input pubkey (to be on the safe side and consistent with the original pubkey extraction for P2PKH inputs).

chimp1984 reviewed (changes requested):
Beside the suggestion to use a block height for activating the change all looks good to me.

stejbac committed:
Exclude segwit pubkeys by block height instead of a blacklist

Since extraction of segwit pubkeys technically represents a hard fork,
activate it by block height in the same way as the fork defined in
TxOutputParser, instead of relying on the absence of premature segwit
BSQ inputs outside of a blacklist. This also means we no longer need to
exclude all but the first tx input from segwit pubkey extraction to
maintain backwards compatibility, which is a little safer and consistent
with the original behaviour of extracting every available P2PKH pubkey.

Provisionally activate this (2nd) DAO hard fork at block height 672646,
which should be 6 weeks from now, just under 5 weeks from the planned
1.5.5 release on 2021/01/27. (Block 1906689 for testnet - 2 weeks from
now assuming an average block time of 10 minutes, but it's erratic.)

stejbac committed:
Ensure RPC client works with Bitcoin Core 0.21.0

Add missing 'connections_(in|out)' JSON properties to NetworkInfo that
came with the recent 0.21.0 release of Bitcoin Core. Also ensure that
unrecognised JSON properties are ignored, so that future changes to the
RPC API are less likely to break our client.

Also, for the benefit of the tests, change the JSON property order of
RawInput to better match observed 'getblock' responses. (It appears that
in 0.21.0, extra "txinwitness" fields have started appearing in coinbase
inputs, which may be a bug.)

chimp1984 reviewed (commented):
utACK

chimp1984 reviewed (commented):
utACK

ripcurlx commented:
@stejbac Could you please resolve the conflict? I hope it will trigger Travis as well (seems to be stuck somehow)

stejbac commented:
I resolved the conflict from the web interface this time. (Perhaps it would have been better to do a rebase - I'm not sure.)

sqrrm commented on core/src/main/java/bisq/core/dao/node/full/rpc/dto/RawBlock.java:
```diff
@@ -0,0 +1,84 @@
+@JsonPropertyOrder({"hash", "confirmations", "strippedsize", "size", "weight", "height", "version", "versionHex",
+        "merkleroot", "tx", "time", "mediantime", "nonce", "bits", "difficulty", "chainwork", "nTx",
+        "previousblockhash", "nextblockhash"})
+public class RawBlock {
```
The name overload of `full/RawBlock` and `full/rpc/dto/RawBlock` is a bit confusing.

stejbac replied:
I just took the names from the very similar DTO classes defined in the original btcd-cli4j library. Perhaps I could rename them from `Raw...` to `DtoBlock`, `DtoTransaction`, etc.

sqrrm replied:
Yeah, I think that would make it clearer, adding a dto prefix or infix for each of the dto classes would make it clear what's going on. It is mostly clear as the variable names you've used include dto, but even better if the class names themselves also adhered to that pattern.

ripcurlx commented:
@stejbac It would be great for this PR to get some information on how to test this best during release testing if it is getting merged for v1.5.5. Thanks!

chimp1984 commented:
@ripcurlx I think it was not intended to get it into 1.5.5 but next release.

stejbac committed:
Rename rpc.dto.RawX to RawDtoX to avoid confusion

Avoid name clashes between the Json RPC client DTOs & the corresponding
raw (un-parsed) DAO state entities. Also prepend the other DTO classes
with 'Dto' for consistency.

stejbac commented:
Rebasing onto a more recent version of master (d3971ef on 2021/01/16, #5072 merge), to avoid the earlier conflict in `RpcService` (caused by a package rename), instead of using a merge. Also adding a commit to address the `RawX` name clash mentioned above.

stejbac committed:
Upgrade jsonrpc4j & Jackson; simplify BitcoindClient

Change jsonrpc4j version from 1.5.3 to 1.6.0.bisq.1, forked to the Bisq
repo from the recent 1.6.0 release. The forked version changes the class
'com.googlecode.jsonrpc4j.HttpException' to be public, instead of (prob.
mistakenly) package private, so we can avoid using reflection to catch
it and re-throw as a 'bisq.network.http.HttpException'. Remove the now
unused constructors from the latter.

As part of this, upgrade Jackson to the latest stable (2.12.1) release,
since jsonrpc4j now depends on a newer version than the previous 2.8.10.

chimp1984 reviewed (dismissed):
utACK

sqrrm commented:
@stejbac Once you remove the draft tag I can merge this PR

chimp1984 commented:
@sqrrm Lets coordinate before merge, we need to test that very well before it goes into master, also considering potential fork risks. Probably best to test it in a dedicated dev branch and only merge quickly before release to reduce time span where master and release have a difference.

sqrrm commented:
@chimp1984 that sounds reasonable. I'm still not able to review this PR but I will do when I'm back in business.

chimp1984 commented:
Activation date is `ACTIVATE_HARD_FORK_2_HEIGHT_MAINNET = 672646;` which would be in about 12 days. I think we need to move that a bit more into the future, depending on release plan. I think most people have updated usually after 2 weeks of release. Though to get extra safety in case there are issues with the release (might be issues not related to that PR) I think a more conservative 4 weeks after release might be better.

@stejbac What would you think?

chimp1984 commented:
@stejbac @ripcurlx @sqrrm @jmacxx 
We need to ensure that all seed nodes as well as explorer nodes have updated ahead in time before release. 
The `ACTIVATE_HARD_FORK_2_HEIGHT_MAINNET` need to be adjusted once release date is more clear.
I tested quite a lot but more testing would be good of course. Not updated nodes partizipating in DAO governance will get invalid issuance proposals in case those have been created with segwit addresses as we have a validation where a LegacyAddress is expected. We should communicate clearly to all DAO stakeholders/contributors that they must update before making proposals or doing voting. I don't see any risk for consensus issues but they would get locally out of sync and would need to resync the DAO state. To reduce risks I think we should try to set the activation date to a non critical period like the proposal period but avoid the more critical ones at voting/blind vote/result.

Thanks @stejbac for the great work!

@ripcurlx Can you arrange the compensation from the remaining funds from the Segwit sponsors? It was agreed that @stejbac will get what remains for that Segwit BSQ implementaion.

stejbac commented:
I think 4 weeks after the 1.5.7 release would be reasonable. Assuming the release is around a week from now, that would place the fork activation in the middle of the Cycle 23 proposal phase, away from the critical periods.

What about the BSQ wallet changes in #5109? Perhaps it would be OK if that PR is merged only into master after the release but before the fork activation, but if #5109 is merged into the release before the fork activation then users will start generating segwit-only BSQ UXTOs, which could get selected as proposal tx inputs. So some DAO participants may encounter problems if they try to make compensation requests early in Cycle 23, as any segwit BSQ tx inputs will be missing the necessary `pubKey` field - I believe this would cause the issued BSQ not to count towards that user's merit as it should (but haven't tested this). Similarly, problems would occur if a user (say the Burning Man) tried to burn any segwit BSQ in a proof-of-burn tx before the fork activation date.

chimp1984 commented:
Ah yes, good point regarding the wallet PR. To me it seems it has to be held back as long the activation time is not triggered. Actually merging this PR asap with short term activation time should be ok as it does not represent much risk. All BSQ txs are still legacy we only support the new RPC and detect pubkeys from segwit txs. After a few weeks when most users have updated we can release #5109 so that segwit addresses will be used.

stejbac commented:
One possible issue with using a short activation time is that anyone running a DAO full node will have monitor hashes out of consensus with the rest of the network if they don't upgrade before the activation date, even though there would be no segwit BSQ txs yet. That's because the segwit BTC (non-BSQ) inputs will also start having nonnull `pubKey` fields after the activation date (though I don't believe that really affects anything other than the hashes).

chimp1984 commented:
Ah ok. So lets play safe with sufficient time for activation and then as second step the wallet release.

Conza88 commented:
General comms being made to prepare users?

stejbac committed:
Activate Segwit DAO hard fork 4 weeks after v1.6.0 release

Push back the (2nd) DAO hard fork activation block height to 680300,
which is 4 weeks after the planned 1.6.0 release around 2021/03/25.
(Also push back the testnet block activation height to 1943000 - 2 weeks
from now assuming an average block time of 10 minutes.)

stejbac committed:
Merge branch 'release/v1.6.0' into implement-segwit-for-bsq

ripcurlx reviewed (approved):
utACK based on https://github.com/bisq-network/bisq/pull/5000#pullrequestreview-583820438

ripcurlx commented:
@stejbac As I'm preparing everything for the v1.6.0 pre-release right now. Is it really necessary that all seednodes are already updated before the pre-release as well? As the activation date is in the future and we don't ship the other Wallet PR yet, it looks to me as if this comment by @chimp1984 is related to the state where this PR contained also the Wallet code, am I wrong?

stejbac commented:
I don't think it's strictly necessary for the seed nodes to be updated at this point (but they should obviously be updated before the hard fork activation date). There shouldn't be any change in the network monitor hashes or any other loss of consensus before that date, unless there is a bug in the code - I didn't spot any such issues during testing on mainnet a month or two ago. I'll run another full DAO rebuild shortly to double check.

stejbac commented:
My full DAO node (running against Bitcoin Core 0.21.0) is still in consensus with the seed nodes after doing a rebuild from genesis. (I'm not sure if there has been an issue with any of the seed nodes recently, though, as my lite node fell out of consensus with them a few days ago for some reason - it wasn't running any of the 1.6.0 changes at the time.)

The current DAO state hash (from my full node) is 44d66d64... @ block height 676797. The proposal and blind vote hashes are also in sync with the seed nodes.

chimp1984 commented:
I can confirm the DAO state hash on my local lite node.

wiz commented:
@chimp1984 @stejbac how should mempool.space display bech32 addresses? currently we add `B` in front: https://mempool.space/bisq/tx/d1f45e55be6101b1b75e6bf9fc5e5341c6ab420647be7555863bbbddd84e92f3

chimp1984 commented:
I think the Bisq UI will handle it the same (B prefix)
